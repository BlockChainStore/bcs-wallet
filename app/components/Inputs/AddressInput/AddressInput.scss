.addressInput {
  .addressItem {
    display: flex;
    align-items: center;
    margin: 0;
    padding: 10px 20px;
    cursor: pointer;

    &:hover {
      background: #eee;
    }

    .meta {
      flex: 1 1 auto;

      .label {
        margin-bottom: 2px;
        color: #444;
        font-size: 14px;
      }

      .value {
        font-size: 12px;
        color: #9e9e9e;
      }
    }

    .actions {
      flex: 0 0 auto;

      .deleteIcon {
        width: 20px;
        height: 20px;
        padding: 8px;
        color: #9BA6B2;
        font-size: 16px;
        cursor: pointer;
      }
    }
  }
}
