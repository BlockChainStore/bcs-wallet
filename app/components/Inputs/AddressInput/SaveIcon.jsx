// @flow
import React from 'react'
import { noop, trim } from 'lodash'

import Tooltip from '../../Tooltip'
import Button from '../../Button'
import Save from 'react-icons/lib/md/save'

import styles from './SaveIcon.scss'

type Props = {
  onSave?: Function
}

type State = {
  saving: boolean,
  name: string
}

export default class AddressInput extends React.Component<Props, State> {
  static defaultProps = {
    onSave: noop
  }

  state = {
    saving: false,
    name: ''
  }

  render = () => {
    return (
      <Tooltip
        className={styles.saveTooltip}
        interactive
        position='left'
        trigger='manual'
        open={this.state.saving}
        html={this.renderTooltip()}
        onRequestClose={this.handleToggle}>
        <Tooltip title='Save to Address Book' position='top'>
          <Save className={styles.saveIcon} tabIndex={0} onClick={this.handleClickIcon} />
        </Tooltip>
      </Tooltip>
    )
  }

  renderTooltip = () => {
    return (
      <div className={styles.saveForm}>
        <label className={styles.label}>Address Name:</label>
        <input
          className={styles.input}
          type='text'
          placeholder='Enter a name for this address'
          value={this.state.name}
          onChange={this.handleChange} />
        <Button
          className={styles.button}
          onClick={this.handleSave}>
          Save to Address Book
        </Button>
      </div>
    )
  }

  handleClickIcon = () => {
    // We need to short-circuit handling the click event on the icon so that it doesn't interfere
    // with the Tooltip onRequestClose prop.  Otherwise, if the user clicks the save icon while the
    // tooltip is already open, then it closes and immediately re-opens, effectively staying open.
    if (!this.state.saving) {
      this.handleToggle()
    }
  }

  handleToggle = () => {
    this.setState({ saving: !this.state.saving, name: '' })
  }

  handleChange = (event: Object) => {
    this.setState({ name: event.target.value })
  }

  handleSave = () => {
    const { onSave } = this.props
    const { name } = this.state
    if (onSave) {
      onSave(trim(name))
    }
  }
}
