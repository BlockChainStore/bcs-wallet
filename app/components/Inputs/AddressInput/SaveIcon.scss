.saveIcon {
  width: 20px;
  height: 20px;
  padding: 8px;
  color: #9BA6B2;
  font-size: 16px;
  cursor: pointer;
}

.saveTooltip {
  pointer-events: auto;
}

.saveForm {
  display: flex;
  flex-direction: column;
  text-align: left;

  .label {
    padding: 10px 0;
  }

  .input {
    width: 250px;
  }

  .button {
    margin: 10px 0;
  }
}
