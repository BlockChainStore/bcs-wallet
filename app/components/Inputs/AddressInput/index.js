// @flow
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import { loadAddresses, saveAddress, deleteAddress, getAddresses } from '../../../modules/addressBook'

import AddressInput from './AddressInput'

const mapStateToProps = (state: Object) => ({
  addresses: getAddresses(state)
})

const actionCreators = {
  loadAddresses,
  saveAddress,
  deleteAddress
}

const mapDispatchToProps = dispatch => bindActionCreators(actionCreators, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(AddressInput)
