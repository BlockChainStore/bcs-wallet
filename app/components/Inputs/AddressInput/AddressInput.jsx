// @flow
import React from 'react'
import Sifter from 'sifter'
import Delete from 'react-icons/lib/md/delete'
import { wallet } from 'neon-js'
import { noop, omit, map, values, trim } from 'lodash'

import SaveIcon from './SaveIcon'
import SelectInput, { DropdownButton } from '../SelectInput'

import styles from './AddressInput.scss'

type ItemType = {
  label: string,
  value: string
}

type Props = {
  value: string,
  addresses?: Object,
  loadAddresses?: Function,
  saveAddress?: Function,
  deleteAddress?: Function
}

export default class AddressInput extends React.Component<Props> {
  static defaultProps = {
    addresses: {},
    loadAddresses: noop,
    saveAddress: noop,
    deleteAddress: noop
  }

  componentDidMount = () => {
    const { loadAddresses } = this.props
    if (loadAddresses) {
      loadAddresses()
    }
  }

  render = () => {
    const passDownProps = omit(this.props, 'addresses', 'loadAddresses', 'saveAddress', 'deleteAddress')

    return (
      <SelectInput
        {...passDownProps}
        className={styles.addressInput}
        items={this.getItems()}
        renderAfter={this.renderAfter}
        renderItem={this.renderItem}
        getItemValue={this.getItemValue}
        getSearchResults={this.getSearchResults} />
    )
  }

  renderAfter = (props: Props) => {
    if (this.canSave()) {
      return <SaveIcon id='saveIcon' onSave={this.handleSave} />
    } else {
      return <DropdownButton {...props} />
    }
  }

  renderItem = (item: ItemType, { onSelect }: { onSelect: Function }) => {
    return (
      <div className={styles.addressItem} key={item.label} tabIndex={0} onClick={onSelect}>
        <div className={styles.meta}>
          <div className={styles.label}>{item.label}</div>
          <div className={styles.value}>{item.value}</div>
        </div>
        <div className={styles.actions}>
          <Delete
            id={`deleteIcon${item.value}`}
            className={styles.deleteIcon}
            tabIndex={0}
            onClick={this.handleDelete(item)} />
        </div>
      </div>
    )
  }

  handleSave = (name: string) => {
    const { saveAddress, value } = this.props
    if (saveAddress) {
      saveAddress(value, name)
    }
  }

  handleDelete = (item: ItemType) => {
    return (event: Object) => {
      event.preventDefault()
      event.stopPropagation()
      const { deleteAddress } = this.props
      if (deleteAddress) {
        deleteAddress(item.label)
      }
    }
  }

  getItems = (): Array<ItemType> => {
    const { addresses } = this.props
    if (!addresses) {
      return []
    }
    return map(addresses, (address, name) => ({
      label: name,
      value: address
    }))
  }

  getItemValue = (item: ItemType): string => {
    return item.value
  }

  getSearchResults = (items: Array<ItemType>, term: string): Array<ItemType> => {
    const sifter = new Sifter(items)
    const result = sifter.search(term, {
      fields: ['label', 'value'],
      sort: [{ field: 'label', direction: 'asc' }]
    })
    return result.items.map((resultItem) => items[resultItem.id])
  }

  canSave = (): boolean => {
    const address = this.props.value
    const { addresses } = this.props

    return trim(address) !== '' && wallet.isAddress(address) && !values(addresses).includes(address)
  }
}
