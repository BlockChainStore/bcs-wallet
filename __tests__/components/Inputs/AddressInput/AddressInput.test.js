import React from 'react'
import { shallow, mount } from 'enzyme'
import AddressInput from '../../../../app/components/Inputs/AddressInput/AddressInput'

describe('AddressInput', () => {
  const props = {
    addresses: {
      'Lenny': 'AW4FD7bz6PF2QadFKF8qXUT7tNmWgvXZc4',
      'Carl': 'AUB7tuEZK63a9fkPEgeeddiZBg7A6PgbZB'
    },
    loadAddresses: jest.fn(),
    saveAddress: jest.fn(),
    deleteAddress: jest.fn()
  }

  test('should render without crashing', () => {
    const wrapper = shallow(<AddressInput {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should load addresses on mount', () => {
    mount(<AddressInput {...props} />)
    expect(props.loadAddresses).toHaveBeenCalled()
  })

  describe('save icon', () => {
    test('should not render by default', () => {
      const wrapper = shallow(<AddressInput {...props} />)
      expect(wrapper.dive().find('#saveIcon').exists()).toEqual(false)
    })

    test('should not render when the value is already saved to the address book', () => {
      const wrapper = shallow(<AddressInput {...props} value='AW4FD7bz6PF2QadFKF8qXUT7tNmWgvXZc4' />)
      expect(wrapper.dive().find('#saveIcon').exists()).toEqual(false)
    })

    test('should render when the value is a valid address', () => {
      const wrapper = shallow(<AddressInput {...props} value='AQzrKoxniN1P4VeP8qyAL4xfX8vJPQ9M8h' />)
      expect(wrapper.dive().find('#saveIcon').exists()).toEqual(true)
    })
  })
})
