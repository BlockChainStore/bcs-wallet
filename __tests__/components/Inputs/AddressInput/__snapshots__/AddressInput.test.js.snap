// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddressInput should render without crashing 1`] = `
<SelectInput
  className="addressInput"
  getItemValue={[Function]}
  getSearchResults={[Function]}
  items={
    Array [
      Object {
        "label": "Lenny",
        "value": "AW4FD7bz6PF2QadFKF8qXUT7tNmWgvXZc4",
      },
      Object {
        "label": "Carl",
        "value": "AUB7tuEZK63a9fkPEgeeddiZBg7A6PgbZB",
      },
    ]
  }
  onChange={[Function]}
  onFocus={[Function]}
  renderAfter={[Function]}
  renderItem={[Function]}
/>
`;
